### ISSUE TITLE
Lint: Temporary lifetime extension extends inaccessible data

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust-clippy#16177]
### What it does

According to [the reference](https://doc.rust-lang.org/1.91.1/reference/destructors.html#r-destructors.scope.lifetime-extension.sub-expressions), when you create and store a reference to a field access expression, a tuple indexing expression, or an indexing expression, lifetime extension can apply to the entire value inside, even though it's no longer accessible.

Clippy should warn when such code is written where something with a destructor is lifetime-extended but becomes inaccessible.

### Advantage

- Run destructor of a value faster, which can avoid unnecessary resource usage or overly-long holding of locks.
- Avoids relying on surpring language semantics.

### Drawbacks

- The lint would suggest changing the code in a way that changes what the code does.
- Could cause code to become more verbose.

### Example

```rust
struct Loud;
impl Drop for Loud {
    fn drop(&mut self) {
        println!("drop");
    }
}

fn make() -> (i32, Loud) {
    (1, Loud)
}

fn main() {
    let x = &make().0;
    println!("{x}");
}
```

Could be written as:

```rust
....

fn main() {
    let temp = make().0;
    let x = &temp;
    println!("{x}");
}
```

Note that the first code prints `1` then `drop`. The second code prints `drop` then `1`.

### Additional Context

There are also some other ways that temporary-lifetime-extended data could become inaccessible. For example:

```rust
....

fn main() {
    let (ref x, _) = make();
    println!("{x}");
}
```

**Comment by danielhenrymantilla** [comment]
Note also that the suggested rewrite is not possible if the wrapping type implements `Drop`, unless the field in question is `Copy`.

(Also, FWIW, the rewrite can be more succintly spelled out as `&{ make().0 }`, although this now becomes an æsthetic preference discussion.)
